### DISCUSSION TITLE
Усреднение рекомендаций по любимым трекам

### DISCUSSION THREAD

**Discussion by Chimildic** [Chimildic/goofy#81] [Личный опыт]
## Описание

Классический пример сбора треков из рекомендованных миксов довольно обширный и разнообразный. Например, [шаблон `Незнакомый сет`](https://chimildic.github.io/goofy/#/template?id=Незнакомый-сет). Если включить сюда миксы жанров, исполнителей и прочее из раздела `Для тебя`, то получается по меньшей мере 300 треков, которые не успеешь прослушать или будешь пропускать большими группами. 

Идея усреднения рекомендаций заключается в следующем:
1. Собрать рекомендованные треки. В первую очередь из миксов дня и подобных. 
2. Применить функцию замены на похожие [replaceWithSimilar](https://chimildic.github.io/goofy/#/func?id=replacewithsimilar) указав в качестве цели для замены недавнюю историю и лайки.
3. Рассчитать среднее значение параметров любимых треков. То есть `tempo` , `energy`, `valence` [и прочие](https://chimildic.github.io/goofy/#/guide?id=Особенности-трека-features).
4. Сделать отклонение от среднего на N% в обе стороны (чтобы получить `min` и `max`)
5. Наложить фильтр [rangeTracks](https://chimildic.github.io/goofy/#/func?id=rangetracks) с аргументами, получившимися на предыдущем шаге. 

То есть опираясь на среднее по любимым трекам, с некоторым отклонением, отбрасываем большую часть рекомендованных треков. И оставляем только те, которые наиболее близки к среднему. 

## Код

Две функции: первая собирает плейлист, вторая реализует фильтр усреднения (шаги 3-5). Скопируйте обе функции к себе и подставьте свои `id` плейлистов-источников.

У фильтра `filterByDeviation_` следующие аргументы:
- (массив) `tracks` - треки, которые попадут в плейлист
- (массив) `templateTracks` - треки, чьи средние значения нужно рассчитать (в примере это любимые треки)
- (число) `percentDeviation` - процент отклонения от среднего. 
- (строка) `keys` - определяет какие ключи учитывать при отборе. Чем больше ключей и меньше процент, тем меньше треков на выходе. Вероятно, лучше всего сочетать какой-либо ключ с темпом `tempo`.

Пример вызова фильтра: среднее по любимым с отклонением в 10% по ключам темпа и энергичности. 
```js
filterByDeviation_(tracks, savedTracks, 10, 'tempo,energy');
```

```js
function filterByDeviation_(tracks, templateTracks, percentDeviation, keys) {
  let cachedFeaturesTracks = getCachedTracks(templateTracks, { features: {} }).features;
  let features = {};

  templateTracks.map(t => t.id).forEach(id => {
    Object.entries(cachedFeaturesTracks[id]).forEach(item => {
      let key = item[0];
      let value = item[1];
      if (!isNaN(value)) {
        features[key] = features[key] ? features[key] + value : value;
      }
    });
  });

  Object.entries(features).forEach(item => {
    let average = item[1] / templateTracks.length;
    features[item[0]] = {
      min: average * (100 - percentDeviation) / 100,
      max: average * (100 + percentDeviation) / 100
    };
  });

  let args = { features: {} };
  keys.split(',').forEach(key => args.features[key] = features[key]);
  Filter.rangeTracks(tracks, args);
}
```

Далее сама функция создания плейлиста, которая использует фильтр
```js
function createDailyPlaylist() {
  let savedTracks = Source.getSavedTracks();
  let tracks = Source.getTracks([
    // Миксы дня
    { id: 'вашеId' },
    // Миксы исполнителя
    { id: 'вашеId' },
    // Миксы жанра
    { id: 'вашеId' },
    // Миксы десятилетия
    { id: 'вашеId' },
  ]);

  Filter.dedupTracks(tracks);
  Filter.replaceWithSimilar({
    origin: tracks,
    replace: RecentTracks.get(3500),
  });
  Filter.matchOriginalOnly(tracks); 
  Filter.removeUnavailable(tracks);
  filterByDeviation_(tracks, savedTracks, 10, 'tempo,energy');

  Playlist.saveWithReplace({
    id: 'вашеId',
    tracks: tracks,
  });
}
```

**Comment by Bondian** [comment]
Спасибо! Очень полезная функция с широким применением. Только вот при подстановке ключа loudness что-то не так срабатывает. Плейлист всегда пустой.

**Comment by ffmusic** [comment]
Incredibly useful to extract songs that you may like from large playlists.
In my specific case, I get better results with:

filterByDeviation_(tracks, savedTracks, 20, 'tempo,energy,valence')

but I suppose this will vary wildly according to the genre of music you might like.

Thanks again, Chilmidic.

**Comment by Chimildic** [comment]
Альтернативный вариант использования фильтра. Вместо усреднения шаблонных треков, эталоном для поиска становится каждый в отдельности. То есть берем от каждого трека его мета данные (energy, valence, danceability и пр) и смотрим на претендентов. Если по всем данным есть совпадение с некоторой погрешностью, одобряем трек. 

При низкой погрешности и большом количестве одновременно сравниваемых мета данных - нужен большой объем входных треков. Иначе выходной результат будет близок к нулю из-за строгости отбора. Объем может обеспечить [mineTracks](https://chimildic.github.io/goofy/#/reference/source?id=minetracks), [craftTracks](https://chimildic.github.io/goofy/#/reference/source?id=crafttracks) или [replaceWithSimilar](https://chimildic.github.io/goofy/#/reference/filter?id=replacewithsimilar).

Побочным эффектом будет большое количество запросов. Поэтому ставить функцию на триггер не рекомендую. Скорее вручную пройтись по интересующим объемам треков.

Таким образом, ядром алгоритма будет следующий код. Где `recomTracks` выходной результат, `poolTracks` массив претендентов, `templateTracks` источник эталонных треков.
```js
  let recomTracks = []
  templateTracks.forEach(template => {
    filterByDeviation_(poolTracks, [template], 5, 'tempo,energy,valence,danceability,acousticness')
    Combiner.push(recomTracks, poolTracks)
    poolTracks = Filter.getLastOutRange()
  })
```

Например, за эталон берем все лайки, крафтим из них рекомендации и фильтруем по мета данным.
```js
function createRelevantPlaylist() {
  let templateTracks = Cache.read('SavedTracks.json')
  let poolTracks = Source.craftTracks(templateTracks)

  let relevantTracks = []
  templateTracks.forEach(template => {
    filterByDeviation_(poolTracks, [template], 5, 'tempo,energy,valence,danceability,acousticness')
    Combiner.push(relevantTracks, poolTracks)
    poolTracks = Filter.getLastOutRange()
  })

  Filter.removeTracks(relevantTracks, templateTracks)

  Playlist.saveWithReplace({
    name: 'Релевантно лайкам',
    tracks: relevantTracks
  })

  Admin.printInfo('Количество запросов', CustomUrlFetchApp.getCountRequest())
}
```

Или майним треки из поиска в большом объеме
```js
  let poolTracks = Source.mineTracks({
    keyword: ['indie', 'alternative'],
    inRow: true,
    itemCount: 50,
    requestCount: 25,
  })
```
